### 01. Steps to classification

1. Build a model.
2. Model learns from the labeled data we pass to it.
3. Pass unlabeled data to the model as input.
4. Model predicts the labels of unseen data. 
* Labeled data = Training data.

### 02. K-Nearest Neighbors

1. Predict the label of a data point by looking at the 'k' closest labeled data points.
2. Majortity vote. 
3. **Question: What is L-Nearest Neighbors?**


     Answer: K-nearest neighbors (K-NN) is a machine learning technique that predicts the class or value of a data point based on the majority class of its nearest neighbors in a training dataset.

### 03. Importing Libraries

In [1]:
from sklearn.neighbors import KNeighborsClassifier

### 04. Reading CSV file

In [2]:
import pandas as pd
churn_df = pd.read_csv("Dataset-1_telecom_churn_clean.csv")
churn_df.head()

Unnamed: 0  account_length  area_code  international_plan  voice_mail_plan  \
0           0             128        415                   0                1   
1           1             107        415                   0                1   
2           2             137        415                   0                0   
3           3              84        408                   1                0   
4           4              75        415                   1                0   

   number_vmail_messages  total_day_minutes  total_day_calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   total_day_charge  total_eve_minutes  total_eve_calls  total_eve_charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   total_night_minutes  total_night_calls  total_night_charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total_intl_minutes  total_intl_calls  total_intl_charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer_service_calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

### 05. Introducing X and y variable

In [5]:
X = churn_df[["total_day_charge","total_eve_charge"]].values
y = churn_df["churn"].values

### 06. Print shapes of X and y

In [6]:
print(X.shape,y.shape)

(3333, 2) (3333,)


(3333,2) means 3333 observations with 2 columns(features)

### 07. Fit the model

In [7]:
knn = KNeighborsClassifier(n_neighbors = 15)
knn.fit(X,y)

KNeighborsClassifier(n_neighbors=15)

### 08. Introducing X_new variable

In [8]:
import numpy as np
X_new = np.array([[56.8,17.5],
                  [24.4,24.1],
                  [50.1,10.9]])
print(X_new.shape)

(3, 2)


### 09. Run the model

In [10]:
predictions = knn.predict(X_new)
print("Predictions: {}".format(predictions))

Predictions: [1 0 0]
